src/klein/_headers.py

```diff
@@ -0,0 +1,356 @@
+
+Any, AnyStr, Iterable, List  # Silence linter
+
+
+__all__ = ()
+
+
+# Interfaces
+
+String = Union[bytes, Text]
+
+RawHeader = Tuple[bytes, bytes]
+RawHeaders = Sequence[RawHeader]
+MutableRawHeaders = MutableSequence[RawHeader]
+
+
+class IHTTPHeaders(Interface):
+    """
+    HTTP entity headers.
+
+    HTTP headers names and values are sort-of-but-not-quite-specifically
+    expected to be text.
+
+    Because the specifications are somewhat vague, and implementations vary in
+    fidelity, both header names and values must be made available as the
+    original bytes received from the network.
+    This interface also makes them available as an ordered sequence of name and
+    value pairs so that they can be iterated in the same order as they were
+    received on the network.
+
+    The C{rawHeaders} attribute provides this as a sequence of C{(name, value)}
+    L{tuple}s.
+
+    A dictionary-like interface that maps text names to an ordered sequence of
+    text values.
+    This assumes that header name bytes are encoded as ASCII, and header value
+    bytes are encoded as ISO-8859-1.
+    """
+
+    rawHeaders = Attribute(
+        """
+        Raw header data as a tuple in the from: C{((name, value), ...)}.
+        C{name} and C{value} are bytes.
+        Headers are provided in the order that they were received.
+        Headers with multiple values are provided as separate name and value
+        pairs.
+        """
+    )  # type: RawHeaders
+
+
+    def getValues(name):
+        # type: (AnyStr) -> Iterable[AnyStr]
+        """
+        Get the values associated with the given header name.
+
+        If the given name is L{bytes}, the value will be returned as the
+        raw header L{bytes}.
+
+        If the given name is L{Text}, the name will be encoded as ASCII and the
+        value will be returned as text, by decoding the raw header value bytes
+        with ISO-8859-1.
+
+        @param name: The name of the header to look for.
+
+        @return: The values of the header with the given name.
+        """
+
+
+
+class IMutableHTTPHeaders(IHTTPHeaders):
+    """
+    Mutable HTTP entity headers.
+    """
+
+    def remove(name):
+        # type: (AnyStr) -> None
+        """
+        Remove all header name/value pairs for the given header name.
+
+        If the given name is L{Text}, it will be encoded as ASCII before
+        comparing to the (L{bytes}) header names.
+
+        @param name: The name of the header to remove.
+        """
+
+
+    def addValue(name, value):
+        # type: (AnyStr, AnyStr) -> None
+        """
+        Add the given header name/value pair.
+
+        If the given name is L{bytes}, the value must also be L{bytes}.
+
+        If the given name is L{Text}, it will be encoded as ASCII, and the
+        value, which must also be L{Text}, will be encoded as ISO-8859-1.
+        """
+
+
+
+# Encoding/decoding header data
+
+HEADER_NAME_ENCODING  = "iso-8859-1"
```

Comment:
@Lukasa: Excellent point…  (/me digs through stale memory…)

This may have been a result of deciding after I wrote the docstrings that using iso-8859-1 was safer because if you are using ASCII as expected, iso-8859-1 will work, and if you are an ASSHAT HTTP JERK that iso-8859-1 will give us a way to capture the CRAPPY BYTES THAT AREN'T HEADER that you sent so that we can at least look at your crap and laugh at it.

Before I fix this one way or the other, how do you feel about just using iso-8859-1 here, and perhaps saying in the docstrings that if you want to be correct, use ASCII?